Resolve #383 -- Map Scripting

cabeca1143 opened the pull request:
*Introduced `IFountain`, `IMapData`, `IMapObject` and `IMapScript` interfaces
*Deleted `IMapProperties`
*Updated a bunch of files to handle the `MapScript` variable instead of the old `MapProperties`.
*Updated `SetLaneID` to allow the change in `laneID` even if it has been already set.
*Removed `SummonersRift.cs` from `GameServerLib/Maps`
*Reworked `Map.cs` and `IMap` in order to house a large ammount of functions for things that were being handled directly in the map's scripts. Also introduced new functions to facilitate the map's scripting and remove it's `Game` variable dependency.
*Introduced `SummonersRiftDefault`, a default map script for vanilla SummonersRift (would be mainly used for maps 2, 6 and 7)
*Introduced a reworked SummonersRift file in `Content/LeagueSandbox-Scripts/Maps` .
*Introduced basic `Howling Abyss` and `Twisted Treeline` scripts
Maps are to be scripted under the `MapScripts` namespace.

cabeca1143 committed:
Transfered from another branch

cabeca1143 committed:
Forgot to Load buildings in summoners rift

cabeca1143 committed:
Unhardcoded Minion Pathing (For maps 10 and 12)

It probably still has to be worked on, since minions will usually stop walking before getting to the the Nexus or Nexus towers, but that's a beginning.
Map1 can't have it's pathing unhardcoded due to the lack of pathing files

cabeca1143 committed:
Some Pathing fixes + Minions wont target protected buildings anymore

danil179 reviewed (commented):
This seems like a change in the right direction. I don't understand the difference between Map1 and SummonersRiftDefault (I only saw two lines for a wrong turret placing or so). In the future I guess we need to unhardcode the many values in the map files into other files (json for example).

cabeca1143 committed:
Missed Tower Change in SummonersRiftDefault

cabeca1143 commented:
> I don't understand the difference between Map1 and SummonersRiftDefault

Now that you said, i think SummonersRiftDefault should get changed to EmptyMapScript and get left the most barebones script possible ,instead of a default SR map. 
I'll change that in a bit.

cabeca1143 committed:
SummonersRiftDefault is now EmptyMapScript, which allows any map to be initialized in a barebones state

cabeca1143 commented:
Done

cabeca1143 committed:
Buildings are now spawned independently from the protection system

This allows the Bypassing of the protection system, which is source of most issues when loading up maps.
The ultimate result is that now ANY map can have at least their towers, inhibitors and nexus spawned, even when using EmptyMapScript

cabeca1143 committed:
Lots of changes in order to properly support Map8 (Crystal Scar)

*Renamed `SpawnObject` to `SpawnLevelProp`
*Map8 now can have it's "towers"(they're actually minions) loaded, not functional tho.
*Fixed Protection system only working with middle inhibs
*There's probably a lot more stuff, but i really can't remember at this point

-All credits related to IRegion/Packets/ILevelProp/ObjectManager changes belong to Lizardy.
Thanks Liz!

lzardy committed:
DebugMode All / IMapData

cabeca1143 committed:
Pull request by microsoftv

DebugMode All / IMapData

lzardy committed:
NotifyS2C_OnEnterTeamVisibility Packet

cabeca1143 committed:
Merge pull request #21 from microsoftv/cabeca-MapStuff

NotifyS2C_OnEnterTeamVisibility Packet

cabeca1143 committed:
Some fixes + Failed Attempt to Spawn to spawn TT's Capture Plates (really close thought)

cabeca1143 committed:
Fixed Typos

lzardy commented on GameServerCore/Maps/IMap.cs:
```diff
@@ -21,25 +25,102 @@ public interface IMap : IUpdate
+        /// <summary>
+        /// List of all nexus in-game
+        /// </summary>
+        List<INexus> _nexus { get; }
```
Underscore is meant for temporary, usually private, variables. Rename this to suit the case, ex: `NexusList`.
The same goes for the rest of the building variables `_turrets`, `_inhibitors`, and `_fountains`.

lzardy commented on GameServerCore/Maps/IMap.cs:
```diff
@@ -21,25 +25,102 @@ public interface IMap : IUpdate
+        /// <summary>
+        /// Spawns Map8's capture points
+        /// </summary>
+        void SpawnCapturePoints();
```
Remove this. This functionality should be handled by the specific map script, not the main one.

lzardy commented on GameServerCore/Maps/IMap.cs:
```diff
@@ -21,25 +25,102 @@ public interface IMap : IUpdate
+        /// <param name="minionNo"></param>
+        /// <param name="barracksName"></param>
+        /// <param name="waypoints"></param>
+        void SpawnLaneMinion(List<MinionSpawnType> list, int minionNo, string barracksName, List<Vector2> waypoints);
```
Rename to CreateLaneMinion. Likely could also change to CreateLaneMinions and setup recursion to handle a full wave of minions.

lzardy commented on GameServerCore/Maps/IMap.cs:
```diff
@@ -21,25 +25,102 @@ public interface IMap : IUpdate
+        /// <param name="waypoints"></param>
+        void SpawnLaneMinion(List<MinionSpawnType> list, int minionNo, string barracksName, List<Vector2> waypoints);
+        IMinion CreateMinion(string name, string model, Vector2 position, uint netId = 0, TeamId team = TeamId.TEAM_NEUTRAL, int skinId = 0, bool ignoreCollision = false, bool isTargetable = false);
+        public void SpawnMinion(IMinion minion);
```
Remove `public`.

lzardy commented on GameServerCore/Scripting/CSharp/IMapScript.cs:
```diff
@@ -0,0 +1,40 @@
+{
+    public interface IMapScript : IUpdate
+    {
+        bool HasInnerTurrets { get; }
```
Remove this and handle each case.

lzardy commented on GameServerCore/Scripting/CSharp/IMapScript.cs:
```diff
@@ -0,0 +1,40 @@
+        long NextSpawnTime { get; set; }
+        long SpawnInterval { get; }
+        //In case someone wishes to use a custom, hardcoded path instead of the one from files
+        bool MinionPathingOverride { get; }
```
Remove this and replace with a check for if MinionPaths variable is empty/not.

lzardy commented on GameServerLib/Config.cs:
```diff
@@ -32,6 +35,7 @@ public class Config
         public bool MinionSpawnsEnabled { get; private set; }
         public string ContentPath { get; private set; }
         public bool IsDamageTextGlobal { get; private set; }
+        //public MapData LoadMapStructures { get; internal set; }
```
Remove this.

lzardy commented on Content/LeagueSandbox-Scripts/Maps/Map8.cs:
```diff
@@ -0,0 +1,354 @@
+
+        public void OnMatchStart()
+        {
+            _map.SpawnCapturePoints();
```
Split this into its individual operations in order to not depend on the main Map script handler.

lzardy commented on GameServerCore/Domain/GameObjects/IInhibitor.cs:
```diff
@@ -6,6 +6,7 @@ public interface IInhibitor : IObjAnimatedBuilding
     {
         bool RespawnAnnounced { get; }
         InhibitorState InhibitorState { get; }
+        public LaneID Lane { get; }
```
Remove `public`.

lzardy commented on GameServerCore/Domain/GameObjects/IMapObject.cs:
```diff
@@ -0,0 +1,21 @@
+{
+    public interface IMapObject
+    {
+        public string Name { get; }
```
Remove 3 `public`s.

lzardy commented on Content/LeagueSandbox-Scripts/Maps/Map10.cs:
```diff
@@ -0,0 +1,393 @@
+        //Dictionary<TeamId, IMinion> BuffPlates = new Dictionary<TeamId, IMinion> { { TeamId.TEAM_BLUE, null }, { TeamId.TEAM_PURPLE, null } };
+        public void OnMatchStart()
+        {   //SPAWNING A MINION WITH THE "TT_Buffplat_L" MODEL BREAKS **ALL** OTHER MINIONS FOR SOME REASON
+
```
Remove all unused comments.

lzardy commented on GameServerCore/Maps/IMap.cs:
```diff
@@ -1,6 +1,10 @@
+using GameServerCore.Enums;
 
 namespace GameServerCore.Maps
 {
```
Rename to `IMapScriptHandler`.

lzardy commented on GameServerLib/Config.cs:
```diff
@@ -5,9 +5,12 @@
 using System.Numerics;
 using System.Reflection;
 using GameServerCore.Domain;
+using GameServerCore.Domain.GameObjects;
```
Remove these 3 newly added `using` lines.
```
using GameServerCore.Domain.GameObjects;
using LeagueSandbox.GameServer.GameObjects.AttackableUnits.AI;
using static LeagueSandbox.GameServer.MapData;
```

lzardy commented on GameServerLib/Config.cs:
```diff
@@ -191,19 +195,23 @@ public GameObjectTypes GetGameObjectType()
                     type = GameObjectTypes.ObjAIBase_Turret;
                 }
-                else if (Name.Contains("__Spawn"))
+                else if (Name.Contains("__Spawn" ))
```
Remove leading space.

lzardy commented on GameServerLib/Config.cs:
```diff
@@ -191,19 +195,23 @@ public GameObjectTypes GetGameObjectType()
                     type = GameObjectTypes.ObjAnimated_HQ;
                 }
-                else if (Name.Contains("Barracks"))
+                else if (Name.Contains("Barracks") && !Name.Contains("Spawn"))
```
Change to `"Barracks_T"` and remove the `&& !Name.Contains("Spawn")`.

lzardy commented on GameServerLib/Config.cs:
```diff
@@ -191,19 +195,23 @@ public GameObjectTypes GetGameObjectType()
                     type = GameObjectTypes.ObjAnimated_BarracksDampener;
                 }
-                else if (Name.Contains("Turret"))
+                else if (Name.Contains("Turret") || Name.Contains("_Point"))
```
Cases where specific map functionality (such as Crystal Scar Neutral Guardian turrets) occurs should be handled by that map script rather than by the map script handler.

Add a separate `else if` for `_Point` and use the `GameObjectTypes.InfoPoint` (you will need to uncomment it in the enum). After, you can setup the turret spawns in the map script.

lzardy commented on GameServerLib/Content/Package.cs:
```diff
@@ -234,6 +235,22 @@ public MapData GetMapData(int mapId)
 
+        //Map12 has different minion spawn names and extra spawns with the old names
+        //So this section checks for the presence of these different files and then removes the unnecessary ones 
+        restart:
```
This whole section goes against guidelines and should be removed and handled somewhere higher level (individual map script preferably).

lzardy commented on GameServerLib/GameObjects/AttackableUnits/AI/LaneMinion.cs:
```diff
@@ -75,7 +75,7 @@ public override bool AIMove()
                 if (!RecalculateAttackPosition())
                 {
-                    KeepFocusingTarget(); // attack/follow target
+                       KeepFocusingTarget(); // attack/follow target
```
Indentation error.

lzardy commented on GameServerLib/GameObjects/AttackableUnits/AttackableUnit.cs:
```diff
@@ -248,8 +248,8 @@ public override void OnRemoved()
-            // We do not want to teleport out of missiles, sectors, or buildings. Buildings in particular are already baked into the Navigation Grid.
-            if (collider is ISpellMissile || collider is ISpellSector || collider is IObjBuilding)
+            // We do not want to teleport out of missiles, sectors, owned regions, or buildings. Buildings in particular are already baked into the Navigation Grid.
+            if (collider is ISpellMissile || collider is ISpellSector || collider is IObjBuilding || (collider is IRegion region && region.CollisionUnit == this))
```
```suggestion
            if (collider is ISpellMissile || collider is ISpellSector || collider is IObjBuilding || (collider is IRegion region && region.CollisionUnit == this || !region.HasCollision))
```

lzardy commented on GameServerLib/GameObjects/Fountain.cs:
```diff
@@ -1,30 +1,31 @@
-    public class Fountain
+    public class Fountain : IFountain
     {
+        private Game _game;
```
Group `private`s separately from `public`s.

lzardy commented on GameServerLib/Maps/Map.cs:
```diff
@@ -1,15 +1,25 @@
+using MapScripts;
 
 namespace LeagueSandbox.GameServer.Maps
 {
```
Rename class (and file name) to `MapScriptManager`.

lzardy commented on GameServerLib/Maps/Map.cs:
```diff
@@ -34,22 +46,36 @@ public class Map : IMap
+        private int _minionNumber;
+        private int _cannonMinionCount;
+        public List<INexus> _nexus { get; set; } = new List<INexus>();
+        public Dictionary<TeamId, IFountain> _fountains { get; set; } = new Dictionary<TeamId, IFountain>();
```
Rename these to reflect `public` (i.e `Fountains`), and so on.

lzardy commented on GameServerLib/Maps/Map.cs:
```diff
@@ -34,22 +46,36 @@ public class Map : IMap
+        private readonly Dictionary<TeamId, SurrenderHandler> _surrenders = new Dictionary<TeamId, SurrenderHandler>();
+        public Dictionary<TeamId, Dictionary<LaneID, List<IInhibitor>>> _inhibitors { get; set; }
+        public Dictionary<TeamId, Dictionary<LaneID, List<ILaneTurret>>> _turrets { get; set; }
+        public List<Tuple<IMinion, uint>> _capturePoints = new List<Tuple<IMinion, uint>>();
```
Remove this, handle via map script.

lzardy commented on GameServerLib/Maps/Map.cs:
```diff
@@ -84,40 +111,487 @@ public void Update(float diff)
+            var inhibRadius = 214;
+            var nexusRadius = 353;
+            var sightRange = 1700;
+            Dictionary<TeamId, List<IMapObject>> test = new Dictionary<TeamId, List<IMapObject>> { { TeamId.TEAM_BLUE, new List<IMapObject>() }, { TeamId.TEAM_PURPLE, new List<IMapObject>() } };
```
Remove this.

lzardy commented on GameServerLib/Maps/Map.cs:
```diff
@@ -84,40 +111,487 @@ public void Update(float diff)
+                {
+                    //Inhibitor model changes dont seem to take effect in-game
+                    _inhibitors[teamId][lane].Add(new Inhibitor(_game, MapScript.InhibitorModels[teamId], lane, teamId, inhibRadius, position, sightRange, Crc32Algorithm.Compute(Encoding.UTF8.GetBytes(mapObject.Name)) | 0xFF000000));
+                    test[teamId].Add(mapObject);
```
Remove this line.

lzardy commented on GameServerLib/Maps/Map.cs:
```diff
@@ -84,40 +111,487 @@ public void Update(float diff)
+                        continue;
+                    }
+
+                    if (mapObject.Name.Contains("_Point"))
```
Change to a separate `else if` using `GameObjectTypes.InfoPoint`.

lzardy commented on GameServerLib/Maps/Map.cs:
```diff
@@ -84,40 +111,487 @@ public void Update(float diff)
+                    _turrets[nexus.Team][LaneID.MIDDLE].FindAll(turret => turret.Type == TurretType.NEXUS_TURRET).ToArray(), TeamInhibitors[nexus.Team].ToArray()
+
+                );
+                var teste = TeamInhibitors[nexus.Team];
```
Remove this line.

lzardy commented on GameServerLib/Maps/Map.cs:
```diff
@@ -84,40 +111,487 @@ public void Update(float diff)
+
+            TurretType inhibProtection = TurretType.INNER_TURRET;
+            //If the map doesn't have Inner turrets, the outer turrets will be used for inhibitor protection instead
+            if (!MapScript.HasInnerTurrets)
```
Replace this check with one which checks for each type of turret.
Inner takes first priority, outer is next, and skip inhib protection when neither are present.

lzardy commented on GameServerLib/Maps/Map.cs:
```diff
@@ -84,40 +111,487 @@ public void Update(float diff)
+            _turrets[team][currentLaneId].Remove(tower);
+            _turrets[team][desiredLaneID].Add(tower);
+        }
+        public void SpawnCapturePoints()
```
Remove this, handle via map script.

lzardy commented on GameServerLib/Maps/Map.cs:
```diff
@@ -84,40 +111,487 @@ public void Update(float diff)
+            return new Tuple<TeamId, Vector2>(teamID, new Vector2(coords.X, coords.Z));
+        }
+
+        public void SpawnLaneMinion(List<MinionSpawnType> list, int minionNo, string barracksName, List<Vector2> waypoints)
```
Change to `CreateLaneMinion`.

lzardy commented on GameServerLib/ObjectManager.cs:
```diff
@@ -76,6 +80,49 @@ public void Update(float diff)
+
+                        // TODO: Implemnent a Region class so we can have a centralized (and cleaner) way of making new regions.
+                        // Turret Vision (values based on packet data, placeholders)
+                        //_game.PacketNotifier.NotifyAddRegion
```
Remove this and replace with a `new Region`.

lzardy commented on GameServerLib/Packets/PacketHandlers/HandleSpawn.cs:
```diff
@@ -68,60 +68,34 @@ public override bool HandlePacket(int userId, SpawnRequest req)
                     if (champion.IsVisibleByTeam(peerInfo.Champion.Team))
                     {
-                        _game.PacketNotifier.NotifyEnterVisibilityClient(champion, userId, true, true, true);
+                        
```
Remove empty space.

lzardy commented on GameServerLib/Packets/PacketHandlers/HandleSpawn.cs:
```diff
@@ -68,60 +68,34 @@ public override bool HandlePacket(int userId, SpawnRequest req)
-
-                        _game.PacketNotifier.NotifyBuyItem((int)turret.NetId, turret, item as IItem);
-                    }
+                    _game.PacketNotifier.NotifyS2C_CreateTurret(turret, userId);
```
Setup `new Region` here as well (always before turret).

lzardy commented on PacketDefinitions420/PacketDefinitions420.csproj:
```diff
@@ -1,17 +1,21 @@
-    <PackageReference Include="ENetSharpLeague" Version="1.2.1-beta" />
-  </ItemGroup>
+	<ItemGroup>
+		<Compile Remove="PacketNotifier-prev.cs" />
```
Remove the `PacketNotifier-prev.cs` file altogether.

lzardy reviewed (changes requested):
Needs a bit of cleaning, otherwise looks good.

cabeca1143 committed:
Requested Changes

lzardy commented on GameServerLib/Maps/Map.cs:
```diff
@@ -84,40 +111,461 @@ public void Update(float diff)
+            }
+
+            //If the map doesn't have any Minion pathing file but the map script has Minion pathing hardcoded
+            if (BlueMinionPathing.Count == 0 && MapScript.MinionPaths != null && MapScript.MinionPaths.Count != 0 || MapScript.MinionPathingOverride)
```
Change this to
```suggestion
            if ((BlueMinionPathing.Count == 0 || MapScript.MinionPathingOverride) && MapScript.MinionPaths != null && MapScript.MinionPaths.Count != 0)
```

cabeca1143 replied:
My intention for the override was that it would actually override everything, .
So if someone wants to do a custom path, they could do even if there were pathing files

lzardy replied:
Conditionals are handled left to right in pairs of two, meaning the original would be `MapScript.MinionPaths.Count != 0 || MapScript.MinionPathingOverride`, the `||` would not be applied to the rest of the conditions before it.

cabeca1143 committed:
More requested changes

lzardy commented on Content/LeagueSandbox-Scripts/Maps/Map8.cs:
```diff
@@ -0,0 +1,353 @@
+
+        public void OnMatchStart()
+        {
+            foreach (var point in _map.CapturePointsList)
```
Change `CapturePointsList` to a `InfoPointList` of mapObjects and handle creation of Minion in-script.

lzardy commented on GameServerCore/Enums/RegionType.cs:
```diff
@@ -0,0 +1,9 @@
+﻿namespace GameServerCore.Enums
+{
+    public enum RegionType
```
Add comment above `// TODO: Rename values to fit function.

lzardy commented on GameServerCore/IGame.cs:
```diff
@@ -25,7 +25,7 @@ public interface IGame: IUpdate
         /// <summary>
         /// Interface for all map properties used for the game.
```
Update summary `Contains all map related game settings such as collision handler, navigation grid, announcer events, and map properties. Doubles as a Handler/Manager for all MapScripts.`

lzardy commented on GameServerCore/Maps/IMapScriptHandler.cs:
```diff
@@ -0,0 +1,130 @@
+namespace GameServerCore.Maps
+{
+    /// <summary>
+    /// Contains all map related game settings such as collision handler, navigation grid, announcer events, and map properties.
```
Update summary `Contains all map related game settings such as collision handler, navigation grid, announcer events, and map properties. Doubles as a Handler/Manager for all MapScripts.`

lzardy commented on GameServerCore/Maps/IMapScriptHandler.cs:
```diff
@@ -0,0 +1,130 @@
+        INavigationGrid NavigationGrid { get; }
+        IMapData MapData { get; }
+        /// <summary>
+        /// MapProperties specific to a Map Id. Contains information about passive gold gen, lane minion spawns, experience to level, etc.
```
Update summary `MapScript which determines functionality for a specific map. Contains information about passive gold gen, lane minion spawns, experience to level, etc.`

lzardy commented on GameServerCore/Maps/IMapScriptHandler.cs:
```diff
@@ -0,0 +1,130 @@
+        /// <summary>
+        /// List of all capture points in-game
+        /// </summary>
+        List<Tuple<IMinion, uint>> CapturePointsList { get; }
```
Change to
```suggestion
        List<Tuple<IMapObject>> InfoPoints { get; }
```

lzardy commented on GameServerLib/Content/Package.cs:
```diff
@@ -202,6 +202,7 @@ public MapData GetMapData(int mapId)
             // SpawnBarracks (lane minion spawn positions)
 
             JObject spawnBarracks = new JObject();
+            bool hasExtraSpawnFiles = false;
```
Remove this unused line.
```suggestion
```

lzardy commented on GameServerLib/Game.cs:
```diff
@@ -104,7 +104,7 @@ public class Game : IGame
         /// <summary>
         /// Interface for all map properties used for the game.
```
Update summary `Contains all map related game settings such as collision handler, navigation grid, announcer events, and map properties. Doubles as a Handler/Manager for all MapScripts.`

lzardy commented on GameServerLib/Maps/MapScriptHandler.cs:
```diff
@@ -0,0 +1,567 @@
+        private readonly Dictionary<TeamId, SurrenderHandler> _surrenders = new Dictionary<TeamId, SurrenderHandler>();
+        public Dictionary<TeamId, Dictionary<LaneID, List<IInhibitor>>> InhibitorList { get; set; }
+        public Dictionary<TeamId, Dictionary<LaneID, List<ILaneTurret>>> TurretList { get; set; }
+        public List<Tuple<IMinion, uint>> CapturePointsList { get; set; } = new List<Tuple<IMinion, uint>>();
```
Change `CapturePointsList` to a `InfoPointList` of mapObjects. Usage of InfoPoints should be handled in-script.
```suggestion
        public List<IMapObject> InfoPoints { get; set; } = new List<IMapObject>();
```

lzardy commented on GameServerLib/Maps/MapScriptHandler.cs:
```diff
@@ -0,0 +1,567 @@
+                }
+                else if (objectType == GameObjectTypes.InfoPoint)
+                {
+                    CapturePointsList.Add(new Tuple<IMinion, uint>(new Minion(_game, null, new Vector2(mapObject.CentralPoint.X, mapObject.CentralPoint.Z), "OdinNeutralGuardian", "OdinNeutralGuardian"), _game.NetworkIdManager.GetNewNetId()));
```
```suggestion
                    InfoPoints.Add(mapObject);
```

lzardy reviewed (changes requested):
Last few changes and it should be good.

lzardy commented on GameServerLib/Maps/MapScriptHandler.cs:
```diff
@@ -0,0 +1,567 @@
+        public IMinion CreateMinion(string name, string model, Vector2 position, uint netId = 0, TeamId team = TeamId.TEAM_NEUTRAL, int skinId = 0, bool ignoreCollision = false, bool isTargetable = false)
+        {
+            var m = new Minion(_game, null, position, model, name, netId, team, skinId, ignoreCollision, isTargetable);
+            _game.PacketNotifier.NotifySpawn(m);
```
```suggestion
            _game.ObjectManager.AddObject(m);
            _game.PacketNotifier.NotifySpawn(m);
```

cabeca1143 committed:
Requested changes

lzardy reviewed (approved):
Good to go.
